x/merkledb/db.go

```diff
@@ -233,53 +226,65 @@ func newDatabase(
 	}
 
 	// mark that the db has not yet been cleanly closed
-	err = trieDB.metadataDB.Put(cleanShutdownKey, didNotHaveCleanShutdown)
+	err = trieDB.baseDB.Put(cleanShutdownKey, didNotHaveCleanShutdown)
 	return trieDB, err
 }
 
 // Deletes every intermediate node and rebuilds them by re-adding every key/value.
 // TODO: make this more efficient by only clearing out the stale portions of the trie.
 func (db *merkleDB) rebuild(ctx context.Context) error {
 	db.root = newNode(nil, RootPath)
-	if err := db.nodeDB.Delete(rootKey); err != nil {
-		return err
-	}
-	it := db.nodeDB.NewIterator()
-	defer it.Release()
 
-	viewSizeLimit := math.Max(
-		db.nodeCache.maxSize/rebuildViewSizeFractionOfCacheSize,
+	opsSizeLimit := math.Max(
+		db.valueNodeDB.nodeCache.maxSize/rebuildViewSizeFractionOfCacheSize,
 		minRebuildViewSizePerCommit,
 	)
-	currentOps := make([]database.BatchOp, 0, viewSizeLimit)
+	intermediateNodeIt := db.baseDB.NewIteratorWithPrefix(intermediateNodePrefixBytes)
+	defer intermediateNodeIt.Release()
+	intermediateBatch := db.baseDB.NewBatch()
+	count := 0
+	// delete all intermediate nodes
+	for intermediateNodeIt.Next() {
+		if count >= opsSizeLimit {
+			if err := intermediateBatch.Write(); err != nil {
+				return err
+			}
+			intermediateBatch = db.baseDB.NewBatch()
+		}
+		count++
+		if err := intermediateBatch.Delete(intermediateNodeIt.Key()); err != nil {
+			return err
+		}
+	}
+	if err := intermediateNodeIt.Error(); err != nil {
+		return err
+	}
+	if err := intermediateBatch.Write(); err != nil {
+		return err
+	}
```

Comment:
We should expand on `database.ClearPrefix` to avoid OOMs and then use that here.

Reply:
Implemented here: #1931